mlir/lib/Tools/PDLL/Parser/Parser.cpp

```diff
@@ -2245,6 +2255,114 @@ FailureOr<ast::Expr *> Parser::parseUnderscoreExpr() {
   return createInlineVariableExpr(type, name, nameLoc, constraints);
 }
 
+FailureOr<ast::Expr *> Parser::parseDictExpr() {
+  consumeToken(Token::l_brace);
+  SMRange loc = curToken.getLoc();
+
+  if (parserContext != ParserContext::Rewrite)
+    return emitError(
+        "Parsing of dictionary attributes as constraint not supported!");
+
+  auto dictAttrCall = createNativeCall(loc, "createDictionaryAttr", {});
+  if (failed(dictAttrCall))
+    return failure();
+
+  // Add each nested attribute to the dict
+  do {
+    FailureOr<ast::NamedAttributeDecl *> decl =
+        parseNamedAttributeDecl(llvm::None);
+    if (failed(decl))
+      return failure();
+
+    ast::NamedAttributeDecl *namedDecl = *decl;
+
+    std::string stringAttrValue =
+        "\"" + std::string((*namedDecl).getName().getName()) + "\"";
+    auto *stringAttr = ast::AttributeExpr::create(ctx, loc, stringAttrValue);
+
+    // Declare it as a variable
+    std::string anonName =
+        llvm::formatv("dict{0}", anonymousDeclNameCounter++).str();
+    FailureOr<ast::VariableDecl *> stringAttrDecl =
+        createVariableDecl(anonName, namedDecl->getLoc(), stringAttr, {});
+    if (failed(stringAttrDecl))
+      return failure();
+
+    // Get its reference
+    auto stringAttrRef = parseDeclRefExpr(
+        (*stringAttrDecl)->getName().getName(), namedDecl->getLoc());
+    if (failed(stringAttrRef))
+      return failure();
+
+    // Create addEntryToDictionaryAttr native call.
+    SmallVector<ast::Expr *> arrayAttrArgs{*dictAttrCall, *stringAttrRef,
+                                           namedDecl->getValue()};
+    auto entryToDictionaryCall =
+        createNativeCall(loc, "addEntryToDictionaryAttr", arrayAttrArgs);
+    if (failed(entryToDictionaryCall))
+      return failure();
+
+    // Uses the new array for the next element.
+    dictAttrCall = entryToDictionaryCall;
+  } while (consumeIf(Token::comma));
+  if (failed(parseToken(Token::r_brace,
+                        "expected `}` to close dictionary attribute")))
+    return failure();
+  return dictAttrCall;
+}
+
+FailureOr<ast::Expr *> Parser::parseArrayAttrExpr() {
+
+  // Advance to the next token without failing.
+  auto nextToken = [&](Token &curToken, int64_t offset) {
+    SMRange loc = curToken.getLoc();
+    SMRange dictArraysLoc(
+        loc.Start.getFromPointer(loc.Start.getPointer() + offset),
+        curToken.getEndLoc());
+    resetToken(dictArraysLoc);
+  };
+
+  SMRange loc = curToken.getLoc();
+
+  const char tokenAfterArray = *loc.End.getPointer();
+  if (tokenAfterArray != '[')
+    return emitError(curToken.getLoc(), "expected `[` after `array`.");
+
+  // Consume `array[` token by advancing 6 characters.
+  // Since the lexer misinterprets `[{` as a string_block, we can't consume the
+  // array token in the normal way. Instead, advance to the next token without
+  // looking at the new Token::Kind.
+  nextToken(curToken, 6);
```

Comment:
After discussing I understand the ambition here was to avoid requiring the user to have a whitespace in `[ {` to disambiguate it from the `[{` token. But we see it is introducing other awkwardnesses if the tokenization is not itself aware that character strings can have multiple interpretations.

Assuming the whitespace between `[ {` does deliver the expected `Token::l_square`, we agree to put the burden on the user for now to add whitespace rather than doing work on context-sensitive tokenization at this time.